compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Error.hs

```diff
@@ -41,6 +42,13 @@ data TemplatePart
   | TPAgreement
   | TPKey
   | TPChoice TemplateChoice
+  | TPInterfaceInstance InterfaceInstanceKey
```

Comment:
I extended `TemplatePart` with a new constructor `TPInterfaceInstance`, mirroring `TPChoice`.

However, even this falls short of the context information I'd like to have: if there's a type error in the body of a method `foo` in `interface instance Bar for Baz` in `template Baz`, the context doesn't mention the method. We could add another constructor to `TemplatePart`, 

```haskell
TPInterfaceInstanceMethod :: InterfaceInstanceKey -> MethodName -> TemplatePart
```

but then `checkInterfaceInstance` would need to construct a `Context` for each method, so it would either need to take a `Context`-making function or the `Template` itself. Even then, we would need to duplicate this for `InterfacePart`.

**tl;dr:** I think a better solution (simpler code and clearer error messages) would be to flatten the `Context` data type (and `TemplatePart`) and change [`Gamma`](https://github.com/digital-asset/daml/blob/main/compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Env.hs#L36-L37) to contain a (perhaps non-empty) stack of `Context`s

Reply:
I let you judge for that, as long you make it in another PR.

Reply:
yes, that's more of a long term idea, I opened https://github.com/digital-asset/daml/issues/14690 so we don't forget